File: src/main/java/edu/cnm/deepdive/Sieve.java
package edu.cnm.deepdive;

import java.util.Arrays;
import java.util.BitSet;

public class Sieve {

  public static int[] sieve(int limit) { // limit is inclusive

    var flags = new BitSet(limit + 1);
    flags.set(2, limit + 1);
    for (var prime = 2; prime <= Math.sqrt(limit); prime = flags.nextSetBit(prime + 1)) {
      for (var multiple = prime * prime; multiple <= limit; multiple += prime) {
        flags.clear(multiple);
      }
    }
    return flags
        .stream()
        .toArray();
  }

  public static void main(String[] args) {
    var upperBound = 100;
    var primes = sieve(upperBound);
    System.out.printf("%d primes from 1 to %d: %s%n", primes.length, upperBound, Arrays.toString(primes));
  }

}

File: src/main/kotlin/edu/cnm/deepdive/sieve.kt
package edu.cnm.deepdive

import kotlin.math.sqrt

fun sieve(limit: Int): List<Int> {
    val primes = mutableListOf<Int>()
    if (limit >= 2) {
        val flags = Array(2) { false } + Array(limit - 1) { true }
        for (value in 2..sqrt(limit.toDouble()).toInt()) {
            if (flags[value]) {
                for (multiple in (value * value)..limit step value) {
                    flags[multiple] = false
                }
            }
        }
        flags.forEachIndexed { index, flag ->
            if (flag) {
                primes.add(index)
            }
        }
    }
    return primes
}

fun main(args: Array<String>) {
    val upperBound = 100
    val primes = sieve(upperBound)
    println("${primes.size} primes found between 1 and $upperBound: $primes")
}